### ISSUE TITLE
Destructible asteroids

### ISSUE THREAD

**Issue by lor1113** [endless-sky/endless-sky#594]
I mean, hello, what kind of asteriod can with stand a quarg skylance for ever??
I think some kind of asteriod HP is needed, maybe make them a ship type with no weapons and just a bit of hull.

**Comment by thenerdfreak** [comment]
The issue with this is that Asteroids are cell based, meaning if someone torches an asteroid with  laser on the other side of the map, one near you would randomly explode

**Comment by bobafetthotmail** [comment]
making them a ship type would do, you would also need a different animation to spawn them (as they can't hyperspace in or take off from planets).

That would also be a basic requirement of any mechanic for asteroid mining.

**Comment by dplepage** [comment]
This is a duplicate of the already-rejected issue #257.

Implementing this is hard because the asteroids are a 4096x4096 grid that repeats endlessly over the system - this is necessary to get a reasonable distribution of asteroids no matter how far from the system center you get without requiring the engine to track infinitely many ships at once.

**Comment by frameworld** [comment]
Maybe track only HP of those you see or saw in the last 5 seconds.

Then make a list sorted by by HP value and spawn a random one not seen from this list when a new one enters your field of view.

Or just 10 steps of 10% and how many asteroids are within that step.
Then use a random step, chance depends on count within each step.
Also assign random HP between step-bottom and step-top values.

**Comment by dplepage** [comment]
@frameworld See #257, where variants on your suggestion have already been discussed. EV did what you suggest - only asteroids near the player actually exist - but this means that missiles following you from a distance won't ever hit asteroids until they get close enough for you to see them, which is weird (and also that distant space battles would be unaffected by asteroids until you got close).

**Comment by frameworld** [comment]
No @dplepage 
Distant asteroids exist, but they only get HP assigned when you either need to see HP or they take damage.

**Comment by bobafetthotmail** [comment]
>"Implementing this is hard because the asteroids are a 4096x4096 grid that repeats endlessly over the system - this is necessary to get a reasonable distribution of asteroids no matter how far from the system center you get without requiring the engine to track infinitely many ships at once."

You cannot do the same but with a ship object instead?

This way only asteroids I can see can also blow up, others I don't see don't exist.

**Comment by dplepage** [comment]
@frameworld I considered trying to implement something like that a little while ago - the asteroids would consist of a base field that repeats infinitely plus a finite list of "exceptions" where asteroids have been damaged or destroyed. The problem you run into is that this list gets way too large very quickly, because you need to add an entry to it every time any asteroid gets hit by anything, which is potentially thousands of asteroids over the course of a long space battle.

@bobafetthotmail No; @endless-sky gave a good explanation for why this doesn't work on #257.

@endless-sky All that said, would it be possible to implement a more EV-like asteroid system by having a coarse, sparse "asteroid probability" field across the system? It'd work something like this:
 * Divide space into a grid of large cells, say 2048x2048, and assign each one an "asteroid density", which is just a single number.
 * Initially, all cells have the same density, which is the system's default density.
 * The densities are stored in a sparse matrix, with entries only being recorded when they differ from the default density (and being removed when they return to the system default).
 * Within some fixed radius of player we track asteroids like ships - they have hit points, can be destroyed, etc.
 * Whenever an asteroid is destroyed, the local density decreases; over time, it'll slowly increase back up to the system's default density (or, if we want to get really fancy, the lower density could diffuse outward along the grid until things normalize).
 * As the player moves, distant asteroids are discard and new ones are created in proportion to the local asteroid density - flying into a zone with low density means you'll see many fewer asteroids.
 * Projectiles moving outside the "real asteroids" radius generate fake asteroid encounters at random, with a frequency proportional to the local density; these lower the local density just like real asteroid destruction.

So we wouldn't actually track distant asteroids, because you can't see them anyway, but you'd still see an approximation of their effects as various projectiles disappear from your map, and you get the realism that the more asteroids get hit by distant weapons fire, the fewer asteroid collisions there will be - if you're approaching a large space battle, you'll probably find near-0 asteroid densities when you get there because all the asteroids have been destroyed by crossfire.

**Comment by knope** [comment]
hmm,
 -- cant we just remove the current method of asteroid placement all together,
 -- change them to ships as previously suggested,
 -- only spawn them out of view (with random trajectories) of the player ship so they seem to float onto the screen.
 -- when you leave the system they do what ever the ships in the system do.

perhaps,
 -- the asteroid can split into smaller fragments (lower HP ships)
 -- the systems can (as they do for pirate/faction systems) have a threshhold for how many to spawn based on where they are on the map?

this game is not taxing as it is, I cant imagine adding a few hundred more ships will make it spike cpu very much...  :)

**Comment by dplepage** [comment]
@knope The problem with your suggestion is this line:

> only spawn them out of view (with random trajectories) of the player ship so they seem to float onto the screen.

@endless-sky wants to avoid that because it means that in an asteroid-dense system, ships firing missiles at you from a distance will never have their missiles hit asteroids until they get very close to you, because asteroids will only exist in a small cloud around your ship. If we create enough asteroid ships that collisions DO happen at any reasonable distance, then we're talking about adding thousands of new ships in some systems, which definitely will overtax things.

What I'm proposing above, though, is pretty much exactly what you're proposing, except with an extra probabilistic component that tries to simulate the effects of an infinite asteroid field on distant ships without actually creating infinitely many asteroids.

**Comment by bobafetthotmail** [comment]
>"@bobafetthotmail No; @endless-sky gave a good explanation for why this doesn't work on #257."

Meh, I was unclear. with "This way only asteroids I can see can also blow up, others I don't see don't exist." I meant something that split the in-view mechanic from the out-of-view mechanic.

In-view is with ship-asteroids, when they go off-view they go poof and disappear. Maybe apart some few you tag for mining or whatever.

The out-of-view mechanic can be a simple RNG roll calculated on the distance target-attacker. For every X kilometers the missile or shot moves through (or must move through total, if you do it in a single routine at launch, then when the shot has to disappear it will disappear) you roll a dice and if the result is Y (depending on system asteroid probability) the missile or shot is stopped by an asteroid.
Continuous weapons need a bit more than this, maybe a flat damage nerf calculated from the system's asteroid probability) but you get the point.

**Comment by frameworld** [comment]
> @frameworld I considered trying to implement something like that a little while ago - the asteroids would consist of a base field that repeats infinitely plus a finite list of "exceptions" where asteroids have been damaged or destroyed. The problem you run into is that this list gets way too large very quickly, because you need to add an entry to it every time any asteroid gets hit by anything, which is potentially thousands of asteroids over the course of a long space battle.

Answered that issue.

List:
 * Elements damaged / looked at
  *  second[1945 modulo 4].list[] = {5 elements}
  *  second[1946 modulo 4].list[] = {27 elements}
  *  second[1947 modulo 4].list[] = {13 elements}
  *  second[1948 modulo 4].list[] = {59 elements}
 * chance[0] = (integer)24 //asteroids with 0% to 10% hp.
 * chance[1] = (integer)12 //asteroids with 10% to 20% hp.
 * chance[2..9] = (integer)(sum: 253) //with 20% to 100% hp.

Asteroid on-damage:
 * is in second[n%4].list[ id ]?
  * yes? delete and insert in the freshest second.list.
  * no? spawn = math.floor( random(0 to 1)*chance[].sum );
  * no? use chance[n=0, n<10, n++] < spawn OR use = use - chance[n].
  * no? remove used from list, insert it to the freshest second.list.

Every second:
 * move oldest second.list into chance[n] dependent on hp (compression with losses of exact hp and position)
 * use second[index] for a new list.

**Comment by endless-sky** [comment]
This idea has been suggested before, but I'm not convinced that it would make the game more enjoyable or add any new interesting strategic elements to the game. The technical issues would be pretty easy to overcome - e.g. asteroids could slowly "heal" so that after a few seconds they go back to being undamaged and thus no longer need to be tracked, and the game could track which "instances" of each asteroid have been damaged to work around the fact that the asteroid field repeats in a tiled pattern. But adding that sort of complexity is pointless if game play is not improved significantly.

**Comment by mdsmestad** [comment]
Your forgeting how fun it is to mindlessly blow up objects. There was this famous old space game..What was it called. Oh ya, ASTEROIDS. Sorry for the sarcasm, but I myself would have hours of fun blowing up asteroids, and scooping up the metals to sell. Maybe you could add some new commoditys for unrefined ore types. Most ore is common and set at a low price, but you could create a chance of finding high class ore on occasion that the player will want to collect.

Expanding on that Idea, I think adding outfits specifically for this task would be great. A mining lazer should be necessary, as well as a outfits like a space scoop or mechanical arms for material collection. 

Also, if you added this dynamic, it would allow you to harvest debree from blown up space ships, that might have a chance to drop outfits and cargo that you could scoop up (though I think the chance should be low, and never allow for the valuable stuff like reactors)

**Comment by frameworld** [comment]
Good ideas @mdsmestad , you are not alone with these.

**Comment by dplepage** [comment]
@endless-sky Part of the benefit would simply be to remove the weird feeling of destroying a bunch of warships with a nuke while these floating rocks around them don't even get nudged off course. Another would be to make "asteroid miner" a viable career option, alongside "trader" and "combat pilot".

That said, I don't think the asteroids currently make a huge impact on the game anyway. Ships aren't maneuverable enough to really take advantage of specific asteroids (e.g. I can't deliberately keep on the opposite side of an asteroid to keep a target from hitting me), so for the most part they just block some fraction of projectiles, making battles happen more slowly. The only serious strategic impact they have is to rule out the specific strategy of running away while dumping lots of homing missiles.

**Comment by dplepage** [comment]
If gameplay were more affected by asteroids, then the ability to destroy them would add a lot to the game. I'd love to see some changes aimed at that, such that flying though a dense asteroid field would be dramatically different from flying in open space. Examples could include:
 * Outfits that become useless in dense asteroids
  * Severely-homing weapons like the needle missiles from EVO, which were low damage but fast and homing, making them almost impossible to dodge (without radar jammers)
  * Low-damage weapons with high-damage submunitions - imagine a missile that travels some distance from your ship and then detonates, sending multiple high-yield warheads in all directions. If it hits an asteroid before it reaches its target distance, the whole thing is destroyed
  * A short-range lightning gun in the style of a tesla coil that can't be aimed but instead always hits the closest target, even if that's an asteroid
  * Highly effective anti-missile systems that can't distinguish between missiles and asteroids and will focus their fire on nearby asteroids even in the face of inbound missiles
 * Outfits that become useful in dense asteroids
  * An array of small tractor beams allowing you to pull in bits of asteroids as ablative armor
  * Smart missiles capable of navigating through asteroid fields
  * A mass-to-fuel converter allowing you to refuel your hyperdrive by scooping up asteroid chunks
  * Lateral thrusters, so that you could actually make use of asteroids when e.g. trying to avoid missiles
 * Other modifications that could make asteroids more interesting
  * Tiny projectile asteroids flitting around, so that you take light shield damage constantly in asteroid-heavy systems
  * Pirate ships/defensive turrets that are disguised as asteroids until you get close to them
  * Asteroid noise on the minimap making it difficult to tell where ships actually were.

With changes like that, the ability to destroy asteroids would add a lot to the game, especially if they were actually pretty hard to destroy (e.g. lasers might do basically nothing to them - unlike a ship where the engine might explode once you punch through the hull, an asteroid with a hole in it isn't much different from an asteroid without one). I can then imagine e.g. weapons that don't damage ships much but allow you to clear asteroids effectively so that you can bring your other weapons into play (or deny your opponents their asteroid-capable advantages).

**Comment by Pointedstick** [comment]
I *LOVE* those ideas, @dplepage!

**Comment by dplepage** [comment]
Actually, I like the tiny asteroids idea so much that I'm going to make a separate issue to discuss it.

**Comment by bobafetthotmail** [comment]
+1 to most.

As they are, asteroids aren't terribly relevant if you don't use missiles.

**Comment by lor1113** [comment]
What's wrong with rendering all asteroids as ships with a hull meter, but that are excluded by targeting with 'r' and 'n'. I'm sure that most computers would be able to handle it, considering it's a 2d game, with not very high-def graphics, nor do the asteroids change course or have any form of AI.
This should be made into a toggleable option for those with powerful enough computers

**Comment by lor1113** [comment]
Besides, it just feels stupid having my quarg skylances blocked by some tiny asteroid, and the mining and base-building aspect is a bonus

**Comment by Thunderforge** [comment]
I don't think that anybody has mentioned yet that Escape Velocity Nova had destroyable asteroids which could be mined for resources. I recall that only certain weapons worked against them and that the destroyed asteroids disappeared after a while. So there is precedent for this and it should be technically feasible if they could do it back in 2002.

**Comment by Pointedstick** [comment]
I mentioned it in https://github.com/endless-sky/endless-sky/issues/257.

**Comment by lor1113** [comment]
If you are worried by performance, why not just reduce the number of asteroids! As you said, you can travel through an asteroid field and not see a single asteroid.

**Comment by endless-sky** [comment]
Cleaning up the issues list to focus on feature requests that are more likely to get implemented.
